UI general update

gfrantzsword opened the pull request:
## :bookmark:	Ticket Link 

N/A

## :interrobang: Type of change

Refactor

## :mag_right: How to validate the PR

3. Add your Cat API key to `local.properties`:
   ```
   API_KEY="your_cat_api_key_here"
   ```
4. Build and run the application

## :open_book: Description

Completely refactored the application UI, with changes to all screens and improved constant organization.

## :camera_flash:	Screenshots
### List Screen (original / updated)
<img width="270" height="600" alt="originalList" src="https://github.com/user-attachments/assets/8d8eb14c-f907-4a94-9170-01048a9d3c0e" />
<img width="270" height="600" alt="updatedList" src="https://github.com/user-attachments/assets/a06b023f-0ae8-42a1-ab1c-3f49717a348e" />

### Favorites Screen (original / updated)
<img width="270" height="600" alt="originalFav" src="https://github.com/user-attachments/assets/b887adc4-e51e-461b-820a-650d131767b9" />
<img width="270" height="600" alt="updatedFav" src="https://github.com/user-attachments/assets/23181566-f16e-46a4-83f4-afc9722796f0" />

### Search (original / updated)
<img width="270" height="600" alt="originalSearch" src="https://github.com/user-attachments/assets/4714d564-ab28-465c-a203-d38ee39a2052" />
<img width="270" height="600" alt="updatedSearch" src="https://github.com/user-attachments/assets/322e1e5c-73e2-4fed-9d9f-16b8bbe5f077" />

### Detail Screen (original / updated)
<img width="270" height="600" alt="originalDetail" src="https://github.com/user-attachments/assets/4ae63864-2fa1-4804-af33-88066a25a32d" />
<img width="270" height="600" alt="updatedDetail" src="https://github.com/user-attachments/assets/dcb3ee26-7602-4f9f-9d8e-5cf71faef14c" />

## :white_check_mark: Checklist

- [x] I have created/updated unit tests
- [X] New and existing unit tests pass locally with my changes
- [ ] I have created documentation

gfrantzsword committed:
fix: general fixes on screen paddings, status-bar and search return

- Adjusted bottom padding on lists to fix issue where elements where hidden behing bottom nav bar;
- Fix Search bar to be singleLine only and return on 'enter';
- Enable edgeToEdge for seamless integration with native status and nav bar

gfrantzsword committed:
feature: Integrate search bar into TopAppBar

Replaces the static text field with a search icon that expands the search bar, creating a cleaner UI

gfrantzsword committed:
refactor: Centralize dimensions, typography and colors

Updated viewModels to use predefined values. This simplifies application-wide UI updates.

gfrantzsword committed:
refactor: Color Palette + minor updates

- Updated the applications color palette
- Add Shadow to TopAppBar

gfrantzsword committed:
refactor: Revamp app theme, colors, and typography

Updated the entire application's visual style to match new design language. This includes a new color palette, updated font styles for titles and text, and standardized shadows and card styles.

gfrantzsword committed:
feature: Add similar breeds to detail screen

Create new 'Similar Breeds' carrousel on breed detail screen
- Displays up to 6 breeds with either the same origin or up at least 3 similar 'temperaments'.

gfrantzsword committed:
feature: Update breeds list to use a 2-column staggered grid
- Adjust spacing for new format
- Fix typo on Dimensions.kt and usage on BreedDetailScreen.kt

gfrantzsword committed:
Fix: Make favorite list reactive

Updated FavoriteDao and Repository to use a Flow. This ensures the
breed list and favorite list update automatically when a favorite is
changed on the detail screen.

gfrantzsword committed:
Refactor: Updated images and typography dimensions values

gfrantzsword committed:
Refactor: Added new innerCornerRadius and ThinBorderEffect to images and updated Dimensions.kt

gfrantzsword committed:
refactor: Add new placeholder and error drawables

Added 2 new custom placeholder and error drawables to replace previous low-res icons

gfrantzsword committed:
fix: Updated screen padding on detail screen

- Updated padding to allow the 'similarBreeds' lazy row to extend the entire view width

gfrantzsword committed:
refactor: Revamped breed-detail UI

- Created StatCards to separate details
- Created TemperamentChips to organize and display multiple 'chips' instead of a string of items

gfrantzsword committed:
refactor: Updated BottomNav icons and color

- Updated breed list to use 'paw' icon instead of 'home';
- Updated selected color to use BrandBlue

AlexCarvalhoSword commented on features/breed_detail/src/main/java/com/example/catbreeds/breed_detail/BreedDetailScreen.kt:
```diff
@@ -86,67 +99,88 @@ fun BreedDetailScreen(
-                            color = MaterialTheme.colorScheme.onSurfaceVariant
-                        )
-                    }
+                    StatCard(
```
try to keep the params names (specially when there are multiple ones). 
Also for the weight, what do you think of creating a Constants file where you can set this recurring values ? So there you will set now the 
`DEFAULT_WEIGHT = 1f`

```suggestion
                    StatCard(
                        modifier =  Modifier.weight(DEFAULT_WEIGHT)
                        label = "Origin",
                        value = breed.origin,
                    )
```

gfrantzsword replied:
Already added the params names. As for the weight:
- I defined the DefaultWeight on my existing Dimensions.kt
- I used UpperCamelCase instead of 'FULL_SNAKE_CASE' to fit with the rest of the constants. If you think I should change that on all of them just let me know, should be a quick refactor

AlexCarvalhoSword commented on features/breed_detail/src/main/java/com/example/catbreeds/breed_detail/BreedDetailScreen.kt:
```diff
@@ -86,67 +99,88 @@ fun BreedDetailScreen(
+                    modifier = Modifier
+                        .fillMaxWidth()
+                        .padding(horizontal = AppDimensions.ScreenPadding),
+                    horizontalArrangement = Arrangement.spacedBy(AppDimensions.InterItemSpacing),
```
You can remove the `AppDimensions` for all the mentions.
```suggestion
                    horizontalArrangement = Arrangement.spacedBy(InterItemSpacing),
```

gfrantzsword replied:
Updated all instances to directly import only the used constants

AlexCarvalhoSword commented on features/breed_detail/src/main/java/com/example/catbreeds/breed_detail/BreedDetailScreen.kt:
```diff
@@ -160,4 +194,106 @@ fun BreedDetailScreen(
+}
+
+@Composable
+fun StatCard(label: String, value: String, modifier: Modifier = Modifier) {
```
can be private like the ones bellow. 
Also, we usually set the modifier as first param

gfrantzsword replied:
Also updated the remaining inner methods across the screen files to be private

AlexCarvalhoSword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/BreedListScreen.kt:
```diff
@@ -53,6 +45,11 @@ fun BreedListScreen(
     val breeds by viewModel.breeds
     val searchQuery by viewModel.searchQuery
     val filteredBreeds by viewModel.filteredBreeds
+    var isSearchActive by remember { mutableStateOf(false) }
```
On our project doing like this is the standard: 
```suggestion
    val isSearchActive = remember { mutableStateOf(false) }
```
Your way is not wrong by any means. Its just a heads up for when you move to our project. You can see the discussion [here](https://swordhealth.slack.com/archives/C02V2U0BDJ5/p1688047888168399)

gfrantzsword replied:
Thanks for the heads up, updated a few similar situations and will keep that in mind

AlexCarvalhoSword commented on features/breed_detail/src/main/java/com/example/catbreeds/breed_detail/BreedDetailScreen.kt:
```diff
@@ -57,11 +50,31 @@ fun BreedDetailScreen(
 
+    // Scroll state for dynamic shadow
+    val scrollState = rememberScrollState()
+    val showTopBarShadow = scrollState.value > 0
```
maybe use `derivedStateOf` here as well

gfrantzsword replied:
I actually used derivedStateOf on the List and Favorites files for the dynamic shadow, this one was just a fluke. Thanks for noticing!

gfrantzsword committed:
refactor: Simplified all 'AppDimensions' mentions

- Directly import each used constant

gfrantzsword committed:
refactor: Define default '1f' weight as a constant

 Defined DefaaultWeight as a constant on Dimensions.kt and directly import it on every instance

gfrantzsword committed:
refactor: Updated inner methods to private and corrected params order on StatCard

- Updated all inner methods across Screen files to private;
- Updated all StatCard method calls on BreedDetailScreen to use actual parameters;
- Updated order of parameters to keep 'modifier' as first

gfrantzsword committed:
refactor: Updated scroll state on BreedDetailScreen and variable definition to be inline with team approach

- Updated the ShowTopBarShadow on BreedDetailScreen to use 'derivedStateOf, similarly to the others Screens
- Adjusted 'var ... by remember {}' to use the team's approach 'val ... = remember {}'

goncalo-guilherme-sword commented on data/src/main/java/com/example/catbreeds/data/repository/BreedRepositoryImpl.kt:
```diff
@@ -70,12 +71,26 @@ class BreedRepositoryImpl(
+            val breed = getBreedById(breedId)
+            if (breed != null) {
+                val allBreeds = getBreeds().first()
+                val temperamentList = breed.temperament.split(", ")
```
I've noticed that you have this logic to split the temperament on multiple places of the code. Maybe you could put this inside of the Breed data class to keep this centralized? Something like this:

`val temperamentList get() = temperament.split(", ")`

joaoreis-sword replied:
IMO this belongs where we do the mapping from BreedEntity to Breed. This is a mapping rule essentially and should live where we do that mapping!

In short `Breed` object should have the temperament field as a `List<String>` and in `BreedEntity.toBreed()` is where we do the mapping.

This is important since we should aim to not tie the implementation of domain objects to database/api models. We should always start by creating the domain objects in a way that makes sense and then do the necessary mapping from the entity/api objects. Imagine that temperament now comes in a different format, we would need to change both BreedEntity and Breed but the domain representation of Breed hasn't change, it's still a list.

joaoreis-sword replied:
To expand even further, I would advise creating `BreedDTO` object that is used for the API requests. This way we're not tying API response details to domain details

gfrantzsword replied:
Thanks for the suggestions 😁 I ended up going for the deeper refactor:
- Created a BreedDto on the network layer;
- Changed Breed.temperament to a `List<String>`;
- Moved the `.split()` logic to BreedEntity.toBreed();
- Updated the mock data and unit tests to work with the new model

I did everything on commit fed1288

goncalo-guilherme-sword commented on features/breed_detail/src/main/java/com/example/catbreeds/breed_detail/BreedDetailScreen.kt:
```diff
@@ -160,4 +206,106 @@ fun BreedDetailScreen(
+}
+
+@Composable
+private fun SimilarBreedCard(
```
Question just to understand your approach. Apart from the image size and the favourite icon, isn't this component the same one that you used on the list?

gfrantzsword replied:
The main differences are essentially the icon, image size and the 'ThinBorderEffect' around the image. I believed those were enough differences to be worth a new method
<img width="102" height="101" alt="image" src="https://github.com/user-attachments/assets/a89aed2e-498d-4f47-9396-acd4214ade76" />
<img width="102" height="101" alt="image" src="https://github.com/user-attachments/assets/0dc52318-2218-432a-9d3a-f0a4ee2b18e3" />

Also I just noticed I'm misusing the Typography on the list's BreedCard!

gfrantzsword committed:
fix: Corrected Type.kt to object

- Refactored The AppTypography definition into an object and correctly import its constants across Screen files

gfrantzsword committed:
Merge branch 'master' into ui-general-update

gfrantzsword committed:
refactor: introduce BreedDto, change breed temperament to list and update tests

- Created 'BreedDto';
- Changed Breed.temperament to a List<String>;
- Moved the temperament '.split(",")' logic to BreedEntity->toBreed();
- Updated tests to represent new model;
- Removed obsolete unit tests from the `app` module.

goncalo-guilherme-sword commented on data/src/main/java/com/example/catbreeds/data/local/BreedEntity.kt:
```diff
@@ -21,7 +21,7 @@ data class BreedEntity(
-            temperament = temperament,
+            temperament = temperament.split(",").map { it.trim() },
             life_span = life_span,
             reference_image_id = reference_image_id,
```
Instead of only having here the image id, I would rename this param to imageUrl and use this param to map here the formatted url of the image? Something like this?
`imageUrl = "https://cdn2.thecatapi.com/images/${reference_image_id}.jpg"`

And then, on the view, when calling the AsyncImage component, you just need to pass `breed.imageUrl` instead of monting the url each time you need it.
What do you think?

gfrantzsword replied:
I originally intended to implement this change on the `add_new_breed` branch but included it here already.

Note: I did **NOT** implement any complex database migration, instead just added a `.fallbackToDestructiveMigration(true)` to handle the schema update.

goncalo-guilherme-sword replied:
You should not call the `fallbackToDestructiveMigration()` because it will nuke the database.
In this case it's "fine" but, on a production environment, that would probably impact all the app users. 
You added that call because the app was crashing right? 
In this case, to not make any migrations, all you needed to do was just uninstall the app and install it again, and it would work just fine,

goncalo-guilherme-sword commented on data/src/main/java/com/example/catbreeds/data/remote/BreedDto.kt:
```diff
@@ -0,0 +1,11 @@
+    val description: String,
+    val temperament: String,
+    val origin: String,
+    val life_span: String,
```
Avoid declare class params with underscore. Since this DTO is used to map the endpoint response, you could use the SerializedName annotation to link the param to the correct param of the response.

`@SerializedName("life_span") val lifeSpan: String`

goncalo-guilherme-sword commented on domain/src/main/java/com/example/catbreeds/domain/models/Breed.kt:
```diff
@@ -4,7 +4,7 @@ data class Breed(
-    val temperament: String,
+    val temperament: List<String>,
     val origin: String,
     val life_span: String,
```
Same here, avoid declaring class params with underscores.

goncalo-guilherme-sword commented on data/src/main/java/com/example/catbreeds/data/local/BreedEntity.kt:
Like on BreedDTO, avoid declare class params with underscores. 
Since this class if for the database, and you are already using the ColumnInfo annotation to set the names on the database columns, just rename the params names and keep the ColumnInfo annotation like it is.

gfrantzsword committed:
refactor: update class parameters to use camelCase

gfrantzsword committed:
refactor: Update breed model to use 'imageUrl'

- Stores the entire url instead of reference id, moving the url construction logic from the view.

gfrantzsword committed:
fix: Added fallbackDestructiveMigration and updated database version

- Included `.fallbackToDestructiveMigration` to automatically drop existing tables to avoid crashes
- Updated database version to from 2 to 3.

gfrantzsword committed:
(PR Suggestion) fix: Remove `.fallbackToDestructiveMigration(true)`

gfrantzsword committed:
Fix: Fix small visual issues on detail screen

- Origin/Life Expectancy row now uses `.height(IntrinsicSize.Max)` to maintain the same height between the cards despite content size;
- Temperament cards now filters empty values;
- Description card now extends full width

These are some issues that could not be seen when using the default API values, but became apparent after the insertion of new local breeds on the `feature/add_new_breed` branch
